subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/AbstractCompiler.java

```diff
@@ -34,18 +35,27 @@
     private final CommandLineToolContext invocationContext;
     private final boolean useCommandFile;
     private final BuildOperationExecutor buildOperationExecutor;
+    private final WorkerLeaseService workerLeaseService;
 
-    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile) {
+    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile, WorkerLeaseService workerLeaseService) {
         this.buildOperationExecutor = buildOperationExecutor;
         this.argsTransformer = argsTransformer;
         this.invocationContext = invocationContext;
         this.useCommandFile = useCommandFile;
         this.commandLineToolInvocationWorker = commandLineToolInvocationWorker;
+        this.workerLeaseService = workerLeaseService;
     }
 
     @Override
-    public WorkResult execute(T spec) {
-        buildOperationExecutor.runAll(commandLineToolInvocationWorker, newInvocationAction(spec));
+    public WorkResult execute(final T spec) {
+        final Action<BuildOperationQueue<CommandLineToolInvocation>> invocationAction = newInvocationAction(spec);
+
+        workerLeaseService.withoutProjectLock(new Runnable() {
+            @Override
+            public void run() {
+                buildOperationExecutor.runAll(commandLineToolInvocationWorker, invocationAction);
```

Comment:
Right. It's a little bit coincidental that every implementation of `newInvocationAction()` does the right thing, but easy to mess up. It would be better to bake this statically into the contract to make it much clearer, but a comment on `newInvicationAction()` would be ok for now. I want to bust this class up anyway, later.